File: backend/app/services/downloader.py
import qbittorrentapi
from abc import ABC, abstractmethod
from loguru import logger
from app.core.config import settings
import asyncio

class DownloaderException(Exception):
    """下载器异常"""
    pass

class BaseDownloader(ABC):
    """下载器基类"""
    
    @abstractmethod
    def add_torrent(self, torrent_url: str, save_path: str = None) -> str:
        """添加种子任务，返回任务ID"""
        pass
    
    @abstractmethod
    def pause_torrent(self, torrent_id: str) -> None:
        """暂停种子任务"""
        pass
    
    @abstractmethod
    def resume_torrent(self, torrent_id: str) -> None:
        """恢复种子任务"""
        pass
    
    @abstractmethod
    def remove_torrent(self, torrent_id: str, remove_files: bool = False) -> None:
        """删除种子任务"""
        pass
    
    @abstractmethod
    def get_torrent_info(self, torrent_id: str) -> dict:
        """获取种子任务信息"""
        pass

class QBittorrentDownloader(BaseDownloader):
    """qBittorrent下载器实现"""
    
    def __init__(self):
        """初始化qBittorrent客户端"""
        try:
            self.client = qbittorrentapi.Client(
                host=settings.DOWNLOADER_HOST,
                username=settings.DOWNLOADER_USERNAME,
                password=settings.DOWNLOADER_PASSWORD
            )
            self.client.auth_log_in()
            logger.info(f"qBittorrent连接成功: {settings.DOWNLOADER_HOST}")
        except Exception as e:

            # logger.error(f"qBittorrent连接失败: {str(e)}")
            raise DownloaderException(f"qBittorrent连接失败: {str(e)}")
    
    async def add_torrent(self, torrent_url: str, save_path: str = None) -> str:
        """添加种子任务"""
        try:
            kwargs = {}
            if save_path:
                kwargs["savepath"] = save_path
            
            self.client.torrents_add(urls=torrent_url, **kwargs)
            
            # 等待片刻，让qBittorrent有时间添加种子
            await asyncio.sleep(2)
            
            # 获取新添加的种子信息以返回ID
            torrents = self.client.torrents_info(sort='added_on', reverse=True, limit=1)
            if torrents:
                return torrents[0].hash
            else:
                raise DownloaderException("无法获取新添加的种子信息")
            
        except Exception as e:
            logger.error(f"添加种子失败: {str(e)}")
            raise DownloaderException(f"添加种子失败: {str(e)}")
    
    def pause_torrent(self, torrent_id: str) -> None:
        """暂停种子任务"""
        try:
            self.client.torrents_pause(torrent_hashes=torrent_id)
        except Exception as e:
            logger.error(f"暂停种子失败: {str(e)}")
            raise DownloaderException(f"暂停种子失败: {str(e)}")
    
    def resume_torrent(self, torrent_id: str) -> None:
        """恢复种子任务"""
        try:
            self.client.torrents_resume(torrent_hashes=torrent_id)
        except Exception as e:
            logger.error(f"恢复种子失败: {str(e)}")
            raise DownloaderException(f"恢复种子失败: {str(e)}")
    
    def remove_torrent(self, torrent_id: str, remove_files: bool = False) -> None:
        """删除种子任务"""
        try:
            self.client.torrents_delete(delete_files=remove_files, torrent_hashes=torrent_id)
        except Exception as e:
            logger.error(f"删除种子失败: {str(e)}")
            raise DownloaderException(f"删除种子失败: {str(e)}")
    
    def get_torrent_info(self, torrent_id: str) -> dict:
        """获取种子任务信息"""
        try:
            torrent = self.client.torrents_info(torrent_hashes=torrent_id)
            
            if not torrent:
                raise DownloaderException(f"未找到种子任务: {torrent_id}")
            
            torrent = torrent[0]
            
            # 计算进度和状态
            progress = torrent.progress
            if torrent.state == "downloading" or torrent.state == "stalledDL":
                status = "downloading"
            elif torrent.state == "pausedDL":
                status = "paused"
            elif torrent.state == "uploading" or torrent.state == "stalledUP" or torrent.state == "pausedUP":
                status = "completed"
            elif torrent.state == "error" or torrent.state == "missingFiles":
                status = "error"
            else:
                status = "downloading"
            
            return {
                "torrent_id": torrent.hash,
                "name": torrent.name,
                "progress": progress,
                "downloaded_bytes": torrent.downloaded,
                "total_bytes": torrent.size,
                "download_speed": torrent.dlspeed,
                "eta": torrent.eta if hasattr(torrent, "eta") else 0,
                "status": status
            }
            
        except Exception as e:
            logger.error(f"获取种子信息失败: {str(e)}")
            raise DownloaderException(f"获取种子信息失败: {str(e)}")

class TransmissionDownloader(BaseDownloader):
    """Transmission下载器实现"""
    
    def __init__(self):
        """初始化Transmission客户端"""
        # 如有需要可以扩展实现Transmission下载器
        raise NotImplementedError("Transmission下载器未实现")
        
    def add_torrent(self, torrent_url: str, save_path: str = None) -> str:
        pass
    
    def pause_torrent(self, torrent_id: str) -> None:
        pass
    
    def resume_torrent(self, torrent_id: str) -> None:
        pass
    
    def remove_torrent(self, torrent_id: str, remove_files: bool = False) -> None:
        pass
    
    def get_torrent_info(self, torrent_id: str) -> dict:
        pass

def get_downloader() -> BaseDownloader:
    """获取下载器实例"""
    downloader_type = settings.DOWNLOADER_TYPE.lower()
    
    if downloader_type == "qbittorrent":
        return QBittorrentDownloader()
    elif downloader_type == "transmission":
        return TransmissionDownloader()
    else:
        logger.error(f"不支持的下载器类型: {downloader_type}")
        raise DownloaderException(f"不支持的下载器类型: {downloader_type}") 

File: backend/internal/service/downloader.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"go.uber.org/zap"

	"github.com/anidog/anidog-go/internal/config"
)

// Downloader 下载器接口
type Downloader interface {
	AddTorrent(ctx context.Context, torrentURL, savePath string) (string, error)
	PauseTorrent(ctx context.Context, torrentID string) error
	ResumeTorrent(ctx context.Context, torrentID string) error
	RemoveTorrent(ctx context.Context, torrentID string, removeFiles bool) error
	GetTorrentInfo(ctx context.Context, torrentID string) (map[string]interface{}, error)
}

// DownloaderException 下载器异常
type DownloaderException struct {
	Message string
}

func (e *DownloaderException) Error() string {
	return e.Message
}

// QBittorrentDownloader qBittorrent 下载器实现
type QBittorrentDownloader struct {
	cfg     *config.Config
	client  *http.Client
	sid     string
	sidMu   sync.Mutex
}

func NewQBittorrentDownloader(cfg *config.Config) *QBittorrentDownloader {
	return &QBittorrentDownloader{
		cfg: cfg,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

func (q *QBittorrentDownloader) ensureLogin(ctx context.Context) error {
	q.sidMu.Lock()
	defer q.sidMu.Unlock()

	if q.sid != "" {
		return nil
	}

	loginURL := strings.TrimRight(q.cfg.DownloaderHost, "/") + "/api/v2/auth/login"
	form := url.Values{
		"username": {q.cfg.DownloaderUsername},
		"password": {q.cfg.DownloaderPassword},
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, loginURL, strings.NewReader(form.Encode()))
	if err != nil {
		return fmt.Errorf("创建登录请求失败: %w", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := q.client.Do(req)
	if err != nil {
		return fmt.Errorf("连接 qBittorrent 失败: %w", err)
	}
	defer resp.Body.Close()

	// 提取 SID cookie
	for _, cookie := range resp.Cookies() {
		if cookie.Name == "SID" {
			q.sid = cookie.Value
			break
		}
	}

	if q.sid == "" {
		return &DownloaderException{Message: "qBittorrent 登录失败：未获取 SID"}
	}

	zap.L().Info("qBittorrent 登录成功")
	return nil
}

func (q *QBittorrentDownloader) apiURL(path string) string {
	return strings.TrimRight(q.cfg.DownloaderHost, "/") + "/api/v2" + path
}

func (q *QBittorrentDownloader) doRequest(ctx context.Context, method, apiPath string, body io.Reader) (*http.Response, error) {
	if err := q.ensureLogin(ctx); err != nil {
		return nil, err
	}

	reqURL := q.apiURL(apiPath)
	req, err := http.NewRequestWithContext(ctx, method, reqURL, body)
	if err != nil {
		return nil, err
	}
	req.AddCookie(&http.Cookie{Name: "SID", Value: q.sid})

	return q.client.Do(req)
}

func (q *QBittorrentDownloader) AddTorrent(ctx context.Context, torrentURL, savePath string) (string, error) {
	form := url.Values{
		"urls":      {torrentURL},
		"savepath":  {savePath},
	}

	resp, err := q.doRequest(ctx, http.MethodPost, "/torrents/add", strings.NewReader(form.Encode()))
	if err != nil {
		return "", &DownloaderException{Message: fmt.Sprintf("添加种子失败: %v", err)}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", &DownloaderException{Message: fmt.Sprintf("添加种子失败，状态码: %d", resp.StatusCode)}
	}

	// 等待种子出现在列表中
	var hash string
	for i := 0; i < 10; i++ {
		time.Sleep(1 * time.Second)
		hash, err = q.findTorrentHash(ctx, torrentURL)
		if err == nil && hash != "" {
			break
		}
	}

	if hash == "" {
		return "", &DownloaderException{Message: "添加种子后未找到对应的 hash"}
	}

	return hash, nil
}

func (q *QBittorrentDownloader) findTorrentHash(ctx context.Context, torrentURL string) (string, error) {
	resp, err := q.doRequest(ctx, http.MethodGet, "/torrents/info", nil)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var torrents []map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&torrents); err != nil {
		return "", err
	}

	for _, t := range torrents {
		if h, ok := t["hash"].(string); ok {
			return h, nil
		}
	}
	return "", fmt.Errorf("未找到种子")
}

func (q *QBittorrentDownloader) PauseTorrent(ctx context.Context, torrentID string) error {
	form := url.Values{"hashes": {torrentID}}
	resp, err := q.doRequest(ctx, http.MethodPost, "/torrents/pause", strings.NewReader(form.Encode()))
	if err != nil {
		return &DownloaderException{Message: fmt.Sprintf("暂停种子失败: %v", err)}
	}
	defer resp.Body.Close()
	return nil
}

func (q *QBittorrentDownloader) ResumeTorrent(ctx context.Context, torrentID string) error {
	form := url.Values{"hashes": {torrentID}}
	resp, err := q.doRequest(ctx, http.MethodPost, "/torrents/resume", strings.NewReader(form.Encode()))
	if err != nil {
		return &DownloaderException{Message: fmt.Sprintf("恢复种子失败: %v", err)}
	}
	defer resp.Body.Close()
	return nil
}

func (q *QBittorrentDownloader) RemoveTorrent(ctx context.Context, torrentID string, removeFiles bool) error {
	form := url.Values{
		"hashes":      {torrentID},
		"deleteFiles": {fmt.Sprintf("%v", removeFiles)},
	}
	resp, err := q.doRequest(ctx, http.MethodPost, "/torrents/delete", strings.NewReader(form.Encode()))
	if err != nil {
		return &DownloaderException{Message: fmt.Sprintf("删除种子失败: %v", err)}
	}
	defer resp.Body.Close()
	return nil
}

func (q *QBittorrentDownloader) GetTorrentInfo(ctx context.Context, torrentID string) (map[string]interface{}, error) {
	resp, err := q.doRequest(ctx, http.MethodGet, fmt.Sprintf("/torrents/info?hashes=%s", torrentID), nil)
	if err != nil {
		return nil, &DownloaderException{Message: fmt.Sprintf("获取种子信息失败: %v", err)}
	}
	defer resp.Body.Close()

	var torrents []map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&torrents); err != nil {
		return nil, err
	}

	if len(torrents) == 0 {
		return nil, &DownloaderException{Message: "种子不存在"}
	}

	t := torrents[0]
	result := make(map[string]interface{})

	if progress, ok := t["progress"].(float64); ok {
		result["progress"] = progress
	}
	if downloaded, ok := t["downloaded"].(float64); ok {
		result["downloaded_bytes"] = int64(downloaded)
	}
	if total, ok := t["size"].(float64); ok {
		result["total_bytes"] = int64(total)
	}
	if speed, ok := t["dlspeed"].(float64); ok {
		result["download_speed"] = int64(speed)
	}
	if eta, ok := t["eta"].(float64); ok {
		result["eta"] = int(eta)
	}

	state := "downloading"
	if s, ok := t["state"].(string); ok {
		switch s {
		case "uploading", "stalledUP", "queuedUP":
			state = "completed"
		case "pausedDL", "pausedUP":
			state = "paused"
		case "error", "missingFiles", "unknown":
			state = "error"
		default:
			state = "downloading"
		}
	}
	result["status"] = state

	return result, nil
}
